File: smartthings_tools/extended_devices/thermostat.py
from __future__ import annotations

from ..base import BaseTool, ToolResult
from ..schemas.extended_devices import ThermostatInput


def _extract_error_message(data: dict) -> str:
    error_detail = data.get("error", {})
    if isinstance(error_detail, dict):
        return error_detail.get("message", str(error_detail))
    return str(error_detail)


class ThermostatTool(BaseTool):
    name = "thermostat_control"
    description = "온도조절기/보일러를 제어합니다. 운전 모드, 난방 목표 온도, 냉방 목표 온도, 팬 모드를 설정할 수 있습니다."
    args_schema = ThermostatInput

    def execute(
        self,
        device_id: str,
        mode: str | None = None,
        heating_setpoint: float | int | None = None,
        cooling_setpoint: float | int | None = None,
        fan_mode: str | None = None,
    ) -> ToolResult:
        commands: list[tuple[str, str, str, list[object], object]] = []

        if mode is not None:
            commands.append(
                ("mode", "thermostatMode", "setThermostatMode", [mode], mode)
            )
        if heating_setpoint is not None:
            commands.append(
                (
                    "heating_setpoint",
                    "thermostatHeatingSetpoint",
                    "setHeatingSetpoint",
                    [heating_setpoint],
                    heating_setpoint,
                )
            )
        if cooling_setpoint is not None:
            commands.append(
                (
                    "cooling_setpoint",
                    "thermostatCoolingSetpoint",
                    "setCoolingSetpoint",
                    [cooling_setpoint],
                    cooling_setpoint,
                )
            )
        if fan_mode is not None:
            commands.append(
                (
                    "fan_mode",
                    "thermostatFanMode",
                    "setThermostatFanMode",
                    [fan_mode],
                    fan_mode,
                )
            )

        if not commands:
            return ToolResult(success=False, error="설정할 항목이 없습니다.")

        applied: dict[str, object] = {}
        responses: dict[str, dict] = {}

        for key, capability, command, arguments, value in commands:
            data = self.client.command(device_id, capability, command, arguments)
            if "error" in data:
                return ToolResult(success=False, error=_extract_error_message(data))
            applied[key] = value
            responses[key] = data

        return ToolResult(
            success=True,
            data={
                "device_id": device_id,
                "set": applied,
                "results": responses,
            },
        )

File: smartthings_tools_rs/src/extended_devices/thermostat.rs
use schemars::schema::RootSchema;
use schemars::schema_for;
use serde_json::Value;

use crate::client::SmartThingsClient;
use crate::common_tools::{extract_error_msg, has_error};
use crate::schemas::extended_devices::ThermostatInput;
use crate::tool::{Tool, ToolResult};

pub struct ThermostatTool<'a> {
    pub client: &'a SmartThingsClient,
}

impl Tool for ThermostatTool<'_> {
    fn name(&self) -> &'static str {
        "thermostat_control"
    }
    fn description(&self) -> &'static str {
        "온도조절기/보일러를 제어합니다. 운전 모드, 난방 목표 온도, 냉방 목표 온도, 팬 모드를 설정할 수 있습니다."
    }
    fn parameters_schema(&self) -> RootSchema {
        schema_for!(ThermostatInput)
    }

    fn execute(&self, args: Value) -> ToolResult {
        let input: ThermostatInput = match serde_json::from_value(args) {
            Ok(v) => v,
            Err(e) => return ToolResult::fail(format!("ValidationError: {e}")),
        };

        let mut commands: Vec<(&str, &str, &str, Vec<Value>)> = Vec::new();

        if let Some(ref mode) = input.mode {
            commands.push((
                "mode",
                "thermostatMode",
                "setThermostatMode",
                vec![Value::String(mode.clone())],
            ));
        }
        if let Some(heat) = input.heating_setpoint {
            commands.push((
                "heating_setpoint",
                "thermostatHeatingSetpoint",
                "setHeatingSetpoint",
                vec![serde_json::json!(heat)],
            ));
        }
        if let Some(cool) = input.cooling_setpoint {
            commands.push((
                "cooling_setpoint",
                "thermostatCoolingSetpoint",
                "setCoolingSetpoint",
                vec![serde_json::json!(cool)],
            ));
        }
        if let Some(ref fan) = input.fan_mode {
            commands.push((
                "fan_mode",
                "thermostatFanMode",
                "setThermostatFanMode",
                vec![Value::String(fan.clone())],
            ));
        }

        if commands.is_empty() {
            return ToolResult::fail("설정할 항목이 없습니다.");
        }

        let mut applied = serde_json::Map::new();
        let mut responses = serde_json::Map::new();

        for (key, capability, command, arguments) in commands {
            let data =
                self.client
                    .command(&input.device_id, capability, command, Some(arguments), None);
            if has_error(&data) {
                return ToolResult::fail(extract_error_msg(&data));
            }
            applied.insert(key.to_string(), serde_json::json!(key));
            responses.insert(
                key.to_string(),
                serde_json::to_value(&data).unwrap_or_default(),
            );
        }

        ToolResult::ok(serde_json::json!({
            "device_id": input.device_id,
            "set": applied,
            "results": responses,
        }))
    }
}
